### PR TITLE
Release candidate 96517

### PR DESCRIPTION
**Pull request by iilyak**
This is based on [01658a0](https://github.com/apache/couchdb/commit/01658a0cc889cffce0ca73b1d06525d708398af8) commit from 2017-11-02.

```
git checkout 01658a0
git checkout -b release-96517
git push cloudant release-96517
git revert d0445f53a837db374e56eb59ede63694ed6e4c33
git revert dae81befd57940d0777402b21c05a2e3c58014d7
# fix conflicts in mango/test/05-index-selection-test.py manually
git revert --continue
git push cloudant release-candidate-96517
```

### CODE REVIEW

File: src/mango/src/mango_cursor.erl [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 create(Db, Selector0, Opts) ->
     Selector = mango_selector:normalize(Selector0),
-    UsableIndexes = mango_idx:get_usable_indexes(Db, Selector, Opts),
+    UsableIndexes = mango_idx:get_usable_indexes(Db, Selector0, Opts),
```
Comment (iilyak) [inline src/mango/src/mango_cursor.erl:49]: This looks suspicious but it was this way before the reverted commit see https://github.com/apache/couchdb/commit/d0445f53a837db374e56eb59ede63694ed6e4c33

**Review by eiri — APPROVED** [review]

**Comment by iilyak** [comment]
It doesn't look like resolved conflicts correctly.
```
..............SSS............................................................SSSSSSSSSSSSSSSSSSSSSSSSSS.....FF.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS....................................SS.......SSSSSSS..........

======================================================================

FAIL: test_uses_all_docs_when_fields_do_not_match_selector (05-index-selection-test.JSONIndexSelectionTests)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/cloudant/couchdb/src/mango/test/05-index-selection-test.py", line 119, in test_uses_all_docs_when_fields_do_not_match_selector

    self.assertNotIn("manager", docs[0])

AssertionError: 'manager' unexpectedly found in {u'ordered': u'aa', u'name': {u'last': u'Hess', u'first': u'Faith'}, u'age': 51, u'email': u'faithhess@pharmex.com', u'manager': True, u'location': {u'city': u'Axis', u'state': u'North Dakota', u'address': {u'street': u'Brightwater Avenue', u'number': 1106}}, u'favorites': [u'Erlang', u'Python', u'Lisp'], u'user_id': 14, u'_id': u'c78c529f-0b07-4947-90a6-d6b7ca81da62', u'company': u'Pharmex', u'_rev': u'1-d954dfeb4831161417bd9ec4c83e0b7f'}

-------------------- >> begin captured logging << --------------------

urllib3.connectionpool: DEBUG: http://127.0.0.1:15984 "POST /mango_test_1b122ad8211e4785a1f42ddd8139b492/_find HTTP/1.1" 200 None

--------------------- >> end captured logging << ---------------------

======================================================================

FAIL: test_uses_all_docs_when_selector_doesnt_require_fields_to_exist (05-index-selection-test.JSONIndexSelectionTests)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/cloudant/couchdb/src/mango/test/05-index-selection-test.py", line 133, in test_uses_all_docs_when_selector_doesnt_require_fields_to_exist

    self.assertEqual(len(docs), 1)

AssertionError: 0 != 1

-------------------- >> begin captured logging << --------------------

urllib3.connectionpool: DEBUG: http://127.0.0.1:15984 "POST /mango_test_1b122ad8211e4785a1f42ddd8139b492/_find HTTP/1.1" 200 None

--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------

Ran 265 tests in 757.535s

FAILED (SKIP=129, failures=2)

make[1]: *** [mango-test] Error 1

make[1]: Leaving directory `/home/travis/build/cloudant/couchdb'

make: *** [check] Error 2
```

**Comment by eiri** [comment]
fwiw the mango tests are passing now:
```
==> couchdb (compile)
./test/build/test-run-couch-for-mango.sh \

[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... ok
[ * ] Running cluster setup ... ok
[ * ] Developers cluster is set up at http://127.0.0.1:15984.
Admin username: testuser
Password: testpass
Time to hack! ... {"couchdb":"Welcome","version":"2.1.1-fa1b0e9","features":["scheduler"],"vendor":{"name":"The Apache Software Foundation"}}
..............SSS............................................................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS....................................SS.......SSSSSSS..........
----------------------------------------------------------------------
Ran 257 tests in 725.178s

OK (SKIP=129)
```

**Comment by iilyak** [comment]
Same here:
```
Time to hack! ... {"couchdb":"Welcome","version":"2.0.0","features":["scheduler"],"vendor":{"name":"The Apache Software Foundation"}}
..............SSS............................................................SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS....................................SS.......SSSSSSS..........
----------------------------------------------------------------------
Ran 257 tests in 724.652s

OK (SKIP=129)
```

**Comment by willholley** [comment]
@iilyak just a note that by default, Mango text index tests will be skipped (because CouchDB doesn't support them). You need to run with environment variable `MANGO_TEXT_INDEXES=1` set to enable them. The CouchDB never runs these because they are essentially Cloudant-specific, so it's important that we run them prior to release. [FB 95200](https://cloudant.fogbugz.com/f/cases/95200/) aims to address this.

**Pull request merged**